import1/dbus.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2019 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package import1 provides integration with the systemd-importd API.  See https://www.freedesktop.org/wiki/Software/systemd/importd/
+// Note: Requires systemd v231 or higher
+package import1
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/godbus/dbus"
+)
+
+const (
+	dbusInterface = "org.freedesktop.import1.Manager"
+	dbusPath      = "/org/freedesktop/import1"
+)
+
+// Conn is a connection to systemds dbus endpoint.
+type Conn struct {
+	conn   *dbus.Conn
+	object dbus.BusObject
+}
+
+// Transfer is the status for an import, export or download operations.
+type Transfer struct {
+	Id       uint32  // The numeric transfer ID of the transfer object
+	Local    string  // The local container name of this transfer
+	Remote   string  // The remote source (in case of download: the URL, in case of import/export a string describing the file descriptor passed in)
+	Type     string  // The type of operation
+	Verify   string  // The selected verification setting, and is only defined for download operations
+	Progress float64 // The current progress of the transfer, as a value between 0.0 and 1.0
+}
+
+// New establishes a connection to the system bus and authenticates.
+// Note: systemd-importd will be activated via D-Bus, we don't need to check service status.
+func New() (*Conn, error) {
+	c := new(Conn)
+
+	if err := c.initConnection(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Conn) initConnection() error {
+	var err error
+	c.conn, err = dbus.SystemBusPrivate()
+	if err != nil {
+		return err
+	}
+
+	// Only use EXTERNAL method, and hardcode the uid (not username)
+	// to avoid a username lookup (which requires a dynamically linked
+	// libc)
+	methods := []dbus.Auth{dbus.AuthExternal(strconv.Itoa(os.Getuid()))}
+
+	err = c.conn.Auth(methods)
+	if err != nil {
+		c.conn.Close()
+		return err
+	}
+
+	err = c.conn.Hello()
+	if err != nil {
+		c.conn.Close()
+		return err
+	}
+
+	c.object = c.conn.Object("org.freedesktop.import1", dbus.ObjectPath(dbusPath))
+
+	return nil
+}
+
+func (c *Conn) getResult(method string, args ...interface{}) (
+	transfer_id uint32, transfer_path dbus.ObjectPath, err error,
```

Comment:
I'd prefer to avoid using named-arguments as return values. Explicit return is a bit easier to read and maintain.

Reply:
OK, I will update it.